pkg/tls/tls.go

```diff
@@ -309,6 +324,54 @@ func getClusterIP(services corev1controllers.ServiceController) (string, error)
 	return service.Spec.ClusterIP, nil
 }
 
+func getHostIPs(deployments appscontrollers.DeploymentController, nodes corev1controllers.NodeController) ([]string, error) {
+	deployment, err := deployments.Get(namespace.System, "rancher", metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	for _, container := range deployment.Spec.Template.Spec.Containers {
+		if container.Name != "rancher" {
+			continue
+		}
+		for _, port := range container.Ports {
+			if port.HostIP != "" || port.HostPort != 0 {
+				return collectNodeIPs(nodes)
+			}
+		}
+	}
+	return nil, nil
+}
+
+func collectNodeIPs(nodeController corev1controllers.NodeController) ([]string, error) {
+	nodes, err := nodeController.List(metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	nodeIPs := make([]string, 0, len(nodes.Items))
+	for _, node := range nodes.Items {
+		var nodeIP string
+		for _, ip := range node.Status.Addresses {
+			if ip.Type == v1.NodeInternalIP {
+				nodeIP = ip.Address
+				break
+			} else if ip.Type == v1.NodeExternalIP && nodeIP == "" {
+				nodeIP = ip.Address
+			}
```

Comment:
Is the idea here to prefer internal? My interpretation of this is that once the internal IP is encountered, that is set/appended, and then we break. I'm not sure that's the correct behavior, as a NodeExternal IP is still valid to reach a node for a hostport service.

Reply:
Yes, you are correct. I will include the external IP as well.